mosaic-runtime/src/main/kotlin/com/jakewharton/mosaic/Static.kt

```diff
@@ -0,0 +1,108 @@
+package com.jakewharton.mosaic
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.ComposeNode
+import androidx.compose.runtime.DisallowComposableCalls
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.mutableStateListOf
+import androidx.compose.runtime.remember
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.collect
+
+/**
+ * Will render each value emitted by [items] as permanent output above the
+ * regular display.
+ */
+@Composable
+fun <T> Static(
+	items: Flow<T>,
+	content: @Composable (T) -> Unit,
+) {
+	class Item(val value: T, var rendered: Boolean)
+
+	// Keep list of items which have not yet been rendered.
+	val pending = remember { mutableStateListOf<Item>() }
+
+	LaunchedEffect(items) {
+		items.collect {
+			pending.add(Item(it, rendered = false))
+		}
+	}
+
+	Static(
+		postRender = {
+			// Remove any items which have been rendered.
+			pending.removeIf { it.rendered }
+		}
+	) {
+		for (item in pending) {
+			Row {
+				// Render item and mark it as having been included in render.
+				content(item.value)
+				item.rendered = true
+			}
+		}
+	}
+}
+
+/**
+ * Renders [content] permanently above the normal canvas. When content has
+ * actually been written to a [TextCanvas], the [postRender] callback will be
+ * invoked to allow clearing of content.
+ *
+ * @param postRender Callback after rendering to a [TextCanvas] is complete.
+ * @param content Content which should be rendered permanently above normal
+ * canvas.
+ */
+@Composable
+internal fun Static(
+	postRender: @DisallowComposableCalls () -> Unit = {},
+	content: @Composable () -> Unit,
+) {
+	ComposeNode<StaticNode, MosaicNodeApplier>(
+		factory = ::StaticNode,
+		update = {
+			set(postRender) {
+				this.postRender = postRender
+			}
+		},
+		content = content,
+	)
+}
+
+internal class StaticNode : ContainerNode() {
+	// Delegate container column for static content.
+	private val box = BoxNode().also {
+		it.isRow = false
+	}
+
+	override val children: MutableList<MosaicNode>
+		get() = box.children
+
+	var postRender: () -> Unit = {}
+
+	override fun measure() {
+		// Not visible.
+	}
+
+	override fun layout() {
+		// Not visible.
+	}
+
+	override fun renderTo(canvas: TextCanvas) {
+		// Render contents of static node to a separate display.
+		val other = box.render()
+
+		// Add display canvas to static canvases if it is not empty.
+		if (other.width > 0 && other.height > 0) {
+			canvas.static.add(other)
+		}
+
+		// Propagate any static content of the display.
+		canvas.static.addAll(other.static)
```

Comment:
This is my least favorite part of the change. The fact that a render causes a side-effect to collect these other renders and then something at the call site magically emits that first.

It's hard to argue with the output, but I think I would like to try and find something better for this. I'm not going to block merging here because the output is truly fantastic and it's all implementation detail.

Render (well, renderTo) is a tree walk, and I think perhaps we separate out the static tree walk to its own function which can be less reliant on side-effects + mutation. It also hopefully means the canvas isn't aware of these static renders.

Reply:
Yeah, this was definitely a result of not wanting to make a bunch of render things `Static` specific or introduce multiple render passes. Thankfully as you said it is all implementation detail so can definitely be iterated.

> Render (well, renderTo) is a tree walk, and I think perhaps we separate out the static tree walk to its own function which can be less reliant on side-effects + mutation.

Are you thinking something like 2 `renderTo` passes, one for static and the other for "normal" content? Or find all the `Static` nodes in the `render` function and render those to the canvas first somehow? I can try a couple different things - if you want - but maybe a little pseudocode of what you are thinking might help.

> It also hopefully means the canvas isn't aware of these static renders.

The tricky part here was that `AnsiOutput` needs to know the static and non-static parts so it can set the `lastHeight` to only the non-static part. So whatever gets sent to the output, needs to know that additional detail.

Reply:
I was thinking a whole separate function, like `renderTo()` and `renderStaticTo()` or something. Basically just splitting the responsibilities and each node would only realistically implement one of them.